### ISSUE TITLE
Progressively fetch package.json

### ISSUE THREAD

**Issue by matthewp** [stealjs/system-npm#41]
Currently we traverse all packages, so we get some that might never be used.  We should be able to get them in normalize when they are dependency of a parent package.

**Comment by olegstepura** [comment]
Definitely should be fixed. A quck tour example bunles 817 lines of package information, most of which is not used. (Actually only 2 packages are used, the `main.js` code and `jquery`)

**Comment by olegstepura** [comment]
Here is the code, that according to a source map is built into a production build file for a single dep (`jqeury`) example on quick start tour: https://gist.github.com/olegstepura/140822352763d1c5a603

**Comment by daffl** [comment]
:+1: Even in smaller projects with three or four dependencies you might end up making 150+ requests to package.json (see http://bitovi-components.github.io/bit-c3/bit-c3/examples/bit-c3.html).

**Comment by justinbmeyer** [comment]
@olegstepura You can use `npmIgnore` until we get this solved.

If anyone wants a crack at solving this, their are two problems to address:

1.  Making builds work.
2.  Keeping extensions like `.stache` working.

## Making builds work

The core problem is that we need `load records`/`nodes` to track every module that is loaded.  We don't treat each `package.json` as an individual file.  Instead, we load them all and combine all their information into a single output of `package.json!npm`.  We do that [here](https://github.com/stealjs/system-npm/blob/master/npm.js#L77).  This looks like what @olegstepura [pasted above](https://gist.github.com/olegstepura/140822352763d1c5a603).

This works great for the build because it simply writes out `package.json!npm`'s output (which is the combination of all `package.json`s).

To fix this, we'd need to be able to dynamically update `package.json!npm` as more and more packages are progressively loaded.  As we get the load record of `package.json`, it might be possible to just update it's `.source` every time we load a new package.

Hopefully the build picks up on this.  If not, it's likely because `.source` isn't intended to be modified after `instantiate`.   If that's the case, we'll have to make some side-table that includes all the sources and progressively modify the source there.  And also make the build use that side table.

## Making extensions like `.stache` work.

I think it's important to be able to say:

```
require("foo.stache!");
```

Without any configuration.  But for this to happen, we have to at least load all the 1st level `package.json`s prior to importing the main module.

**Comment by matthewp** [comment]
@justinbmeyer With system-trace we now keep each module's load object and you can modify it with

```js
System.getModuleLoad("package.json!npm").source = "new source";
```

This should be the same object retained [by the build](https://github.com/stealjs/steal-tools/blob/master/lib/graph/make_graph.js#L81).

**Comment by blink1073** [comment]
Would you be amenable to storing the *full* `package.json` `Object` for the top level modules under `System.npm`, as opposed to the current stub?  We end up re-fetching the top level `package.json` files to get pertinent information in our app.

**Comment by justinbmeyer** [comment]
@blink1073 this is a related, but different issue.  Can you create another issue and we will discuss it there?  Thanks!

**Comment by whitecolor** [comment]
What so difficult with it? Probably that effects loading time much.

**Comment by matthewp** [comment]
I have this partially working in a branch, including overwriting the `package.json!npm` source as packages are progressively loaded.

Current issue is with config. For example canjs has a config like:

```json
"map": {
  "jquery/jquery": "jquery"
}
```

Which we convert to:

```json
"map": {
  "jquery@0.0.0#jquery": "jquery@0.0.0#index"
}
```

This won't work when progressively fetching because we don't have jquery's package.json and so can't convert it.

What I'd like to do instead is when we encounter a package dependency that we cannot convert we store that into a side-table. When jquery's package.json is later loaded in normalize we then convert and apply the config.

Note that we'll need to reapply the `package.json!npm` source at that time as well.

**Comment by justinbmeyer** [comment]
> we don't have jquery's package.json and so can't convert it.

Currently, this conversion happens after all package.json's have been loaded?

**Comment by matthewp** [comment]
Yes.

**Comment by matthewp** [comment]
This has been merged, will land in Steal 0.14

**Comment by whitecolor** [comment]
@matthewp cool
